kernel/src/main/scala/cats/kernel/Hash.scala

```diff
@@ -0,0 +1,52 @@
+package cats.kernel
+
+import scala.{specialized => sp}
+import scala.util.hashing.Hashing
+
+/**
+ * A type class used to represent a hashing scheme for objects of a given type.
+ * For any two instances `x` and `y` that are considered equivalent under the
+ * equivalence relation defined by this object, `hash(x)` should equal `hash(y)`.
+ * @author Tongfei Chen
+ */
+trait Hash[@sp A] extends Any with Eq[A] with Serializable { self =>
+
+  /**
+   * Returns the hash code of the given object under this hashing scheme.
+   */
+  def hash(x: A): Int
```

Comment:
consider using `Long` and move away from the JVM legacy?

Reply:
I disagree about using `Int`. If you have a `Long` you can easily get an `Int` from int, but not the other way around. Some nice algorithms using hashes want more than 32 bits. For instance hyperloglog with only 32 bits would not be great.

Reply:
+1 to using `Long` here. If you want to just give me an `Int` upcast to `Long` that's on you, but it would be better to enable better algorithms.